tests/test_copy.py

```diff
@@ -353,3 +355,239 @@ def __fspath__(self):
+                'copytab', source=f, columns=('a', 'b~'),
+                schema_name='public', format='csv',
+                delimiter='|', null='n-u-l-l', header=True,
+                quote='*', escape='!', force_not_null=('a',),
+                force_null=('b~',))
+
+            self.assertEqual(res, 'COPY 7')
+
+            await self.con.execute('SET search_path=public')
+
+            output = await self.con.fetch("""
+                SELECT * FROM copytab ORDER BY a
+            """)
+            self.assertEqual(
+                output,
+                [
+                    ('*', None, None),
+                    ('a1', 'b1', None),
+                    ('a2', 'b2', None),
+                    ('a3', 'b3', None),
+                    ('a4', 'b4', None),
+                    ('a5', 'b5', None),
+                    ('n-u-l-l', 'bb', None),
+                ]
+            )
+
+        finally:
+            await self.con.execute('DROP TABLE public.copytab')
+
+    async def test_copy_to_table_large_rows(self):
+        await self.con.execute('''
+            CREATE TABLE copytab(a text, b text);
+        ''')
+
+        try:
+            class _Source:
+                def __init__(self):
+                    self.rowcount = 0
+
+                @compat.aiter_compat
+                def __aiter__(self):
+                    return self
+
+                async def __anext__(self):
+                    if self.rowcount >= 100:
+                        raise StopAsyncIteration
+                    else:
+                        self.rowcount += 1
+                        return b'a1' * 500000 + b'\t' + b'b1' * 500000 + b'\n'
+
+            res = await self.con.copy_to_table('copytab', source=_Source())
+
+            self.assertEqual(res, 'COPY 100')
+
+        finally:
+            await self.con.execute('DROP TABLE copytab')
+
+    async def test_copy_to_table_from_bytes_like(self):
+        await self.con.execute('''
+            CREATE TABLE copytab(a text, b text);
+        ''')
+
+        try:
+            data = memoryview((b'a1' * 500 + b'\t' + b'b1' * 500 + b'\n') * 2)
+            res = await self.con.copy_to_table('copytab', source=data)
+            self.assertEqual(res, 'COPY 2')
+        finally:
+            await self.con.execute('DROP TABLE copytab')
+
+    async def test_copy_to_table_fail_in_source_1(self):
+        await self.con.execute('''
+            CREATE TABLE copytab(a text, b text);
+        ''')
+
+        try:
+            class _Source:
+                def __init__(self):
+                    self.rowcount = 0
+
+                @compat.aiter_compat
+                def __aiter__(self):
+                    return self
+
+                async def __anext__(self):
+                    raise RuntimeError('failure in source')
+
+            with self.assertRaisesRegexp(RuntimeError, 'failure in source'):
+                await self.con.copy_to_table('copytab', source=_Source())
```

Comment:
Can we add ```self.assertEqual(await self.con.fetchval('select 1'), 1)``` to all tests that test failures to make sure that the protocol is fully recovered?